chalk-ir/src/lib.rs

```diff
@@ -2629,6 +2630,79 @@ pub enum Constraint<I: Interner> {
 
 impl<I: Interner> Copy for Constraint<I> where I::InternedLifetime: Copy {}
 
+/// A list of constraints.
+#[derive(Copy, Clone, PartialEq, Eq, Hash, HasInterner)]
+pub struct Constraints<I: Interner> {
+    interned: I::InternedConstraints,
+}
+
+impl<I: Interner> Constraints<I> {
+    /// Creates a list of constraints.
+    pub fn from(
+        interner: &I,
+        constraints: impl IntoIterator<Item = impl CastTo<InEnvironment<Constraint<I>>>>,
+    ) -> Self {
+        Self::from_fallible(
+            interner,
+            constraints
+                .into_iter()
+                .map(|p| -> Result<InEnvironment<Constraint<I>>, ()> { Ok(p.cast(interner)) }),
+        )
+        .unwrap()
+    }
+
+    /// Try to create a list of constraints.
+    pub fn from_fallible<E>(
+        interner: &I,
+        constraints: impl IntoIterator<Item = Result<impl CastTo<InEnvironment<Constraint<I>>>, E>>,
+    ) -> Result<Self, E> {
+        use crate::cast::Caster;
+        Ok(Constraints {
+            interned: I::intern_constraints(interner, constraints.into_iter().casted(interner))?,
+        })
+    }
+
+    /// Get the interned representation of the constraints.
+    pub fn interned(&self) -> &I::InternedConstraints {
+        &self.interned
+    }
+
+    /// Index into the list of constraints.
```

Comment:
We gotta find a way to remove this boilerplate

Reply:
Definitely agree, we have basically this exact code duplicated for at least 4 other types. I'll open an issue for it.